x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,222 @@
+		slots[i].Reqs = reqMap
+	}
+
+	return slots
+}
+
+// group the slots
+func GroupSlots(slots []*PairingSlot) []*PairingSlot {
+	uniqueSlots := []*PairingSlot{}
+
+	if len(slots) == 0 {
+		panic("no pairing slots available")
+	}
+
+	uniqueSlots = append(uniqueSlots, slots[0])
+	for k := 1; k < len(slots); k++ {
+		isUnique := true
+
+		for i := range uniqueSlots {
+			if slots[k].Equal(uniqueSlots[i]) {
+				uniqueSlots[i].Count += 1
+				isUnique = false
+				break
+			}
+		}
+
+		if isUnique {
+			uniqueSlots = append(uniqueSlots, slots[k])
+		}
+	}
+
+	return uniqueSlots
+}
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func GetStrategy() ScoreStrategy {
+	return uniformStrategy
+}
+
+// CalcPairingScore calculates the final pairing score for a pairing slot (with strategy)
+// For efficiency purposes, we calculate the score on a diff slot which represents the diff reqs of the current slot
+// and the previous slot
+func CalcPairingScore(scores []*PairingScore, strategy ScoreStrategy, diffSlot *PairingSlot, minStake sdk.Int) error {
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight := strategy[reqName] // not checking if key is found because it's verified in init()
+
+		for _, score := range scores {
+			newScoreComp := req.Score(*score.Provider)
+			if newScoreComp == 0 {
+				err := fmt.Errorf("score component is zero. score component name: %s, provider address: %s", reqName, score.Provider.Address)
+				panic(err)
+			}
+			newScoreComp = commontypes.SafePow(newScoreComp, weight)
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				score.Score /= prevReqScoreComp
+			}
+			score.Score *= newScoreComp
+
+			// update the score component map
+			score.ScoreComponents[reqName] = newScoreComp
+		}
+	}
+
+	return nil
+}
+
+// PrepareHashData prepares the hash needed in the pseudo-random choice of providers
+func PrepareHashData(projectIndex string, chainID string, epochHash []byte) []byte {
+	hashData := []byte{}
+
+	// add the session start block hash to the function to make it as unpredictable as we can
+	hashData = append(hashData, epochHash...)
+	hashData = append(hashData, chainID...)      // to make this pairing unique per chainID
+	hashData = append(hashData, projectIndex...) // to make this pairing unique per consumer
+
+	return hashData
+}
+
+// PickProviders pick a <group-count> providers set with a pseudo-random weighted choice (using the providers' score list and hashData)
+func PickProviders(ctx sdk.Context, scores []*PairingScore, groupCount int, hashData []byte, chosenProvidersIdx map[int]bool) (returnedProviders []epochstoragetypes.StakeEntry) {
+	if len(scores) == 0 {
+		return returnedProviders
+	}
+
+	var scoreSum uint64
+	for idx, providerScore := range scores {
+		if chosenProvidersIdx[idx] {
```

Comment:
this implementation uses extra overhead and keeps updating these providers scores even if they are no longer relevant
we can save this extra memory and overhead, by setting an empryScoreObject inside the slots of already chosen providers, with score 0, this way we will also not update them when updating score

Reply:
done. see https://github.com/lavanet/lava/commit/9fdaee7f74bb1c371aebabd65334ee396226e6f9

Reply:
to be fixed

Reply:
done. see https://github.com/lavanet/lava/commit/6a09f8a2ab74ad252278dc19090569bb1cd4d889